wordpress-mobile/WordPress-Android
# [Dependency Updates] Update `kotlinxCoroutinesVersion` to 1.6.4

Parent #17560
Unblocks: #17561

This PR updates `kotlinxCoroutinesVersion` to [1.6.4](https://github.com/Kotlin/kotlinx.coroutines/releases/tag/1.6.4).

Also, as part of this update the below transitive dependencies were added on the `WordPress` module (4c3fa547264661ecd74ce6116c0092e493fe0f9d):

- `org.jetbrains.kotlinx:kotlinx-coroutines-core`

PS: I recommend doing the code review process commit-by-commit and with the help of AS (or any IDE for that matter). This is because some commits have hundreds of one line changes, and that, might be overwhelming to do via GitHub alone.

-----

Migration Guide
- [README.md](https://github.com/Kotlin/kotlinx.coroutines/blob/1.6.0/kotlinx-coroutines-test/README.md)
- [MIGRATION.md](https://github.com/Kotlin/kotlinx.coroutines/blob/1.6.0/kotlinx-coroutines-test/MIGRATION.md)

FYI: If it helps, WCAndroid did such an update a while ago (`Apr 19 2022`). As such, if you would like to reference or cross-check changes, feel free to visit that update here: [Bump coroutines version to 1.6 #6292](https://github.com/woocommerce/woocommerce-android/pull/6292)

-----

Migration/Structural Changes:

1. [Rename main coroutine scope rule to coroutine test rule.](https://github.com/wordpress-mobile/WordPress-Android/commit/4b9c30dfd68fbe3febfd60d02e5ea6e4fdbe8b08)
2. [Migrate to new coroutine test rule.](https://github.com/wordpress-mobile/WordPress-Android/commit/e6f3f89e9ad8d7496be8e4779c8ad70fe7ada40d)
1. [Replace coroutines utils test utility function with base unit test.](https://github.com/wordpress-mobile/WordPress-Android/commit/3d1e478c0f4043023f1062c7841b4326f5438e2b)
2. [Replace instant task executor rule function with base unit test.](https://github.com/wordpress-mobile/WordPress-Android/commit/5ed5e2c255b0017b59a0609bf7a5ecb29d30cff0)
3. [Replace coroutines utils with coroutines test rule test dispatcher.](https://github.com/wordpress-mobile/WordPress-Android/commit/290466a5500b56b0d368f9524ad0830bc5f95879) (`CoroutinesUtils.TEST_DISPATCHER` replacement/removal)
4. [Replace run blocking test with test from base unit test.](https://github.com/wordpress-mobile/WordPress-Android/commit/c4cf6855470d070e22d235b3d2ea902bc01ad264)
5. [Replace coroutines utils with coroutines test rule test scope.](https://github.com/wordpress-mobile/WordPress-Android/commit/3035d4da6bc5330cd2eb9c90f59db905196af27a) (`CoroutinesUtils.testScope()` replacement/removal)
6. [Remove the now unused coroutines utils file.](https://github.com/wordpress-mobile/WordPress-Android/commit/54ff7ef89c5e6d6b0cc2a603b8281276125b0de0) (`CoroutinesUtils` removal)
7. [Replace custom dispatcher with test dispatcher from base unit test.](https://github.com/wordpress-mobile/WordPress-Android/commit/659ebdd62932577e1fee8a09c1317a32f804db9d)

Fix Tests Changes:

1. [Fix coroutines related test failure due to field test class init.](https://github.com/wordpress-mobile/WordPress-Android/commit/208bfbfc14dc398abf6c88a85df95124c784e38e)
2. [Fix use case load data tests with advance until idle.](https://github.com/wordpress-mobile/WordPress-Android/commit/f3ae6d8eb629f6514fa6394fe1f03b2497586f9e)
3. [Fix assert view all followers second load test condition.](https://github.com/wordpress-mobile/WordPress-Android/commit/515866508eb9a3b8f521ec72a00741912bbecb25)
4. [Fix maps full most popular insights to ui model test condition.](https://github.com/wordpress-mobile/WordPress-Android/commit/ff3d9e8120036d4389127309c6ec5c6a0e5b9fdc)
5. [Fix maps referrers to ui model test conditions.](https://github.com/wordpress-mobile/WordPress-Android/commit/19f4318de5f28801cd8e9cc4e588dbd1c33b829c)
6. [Fix load latest post summary tests with advance until idle.](https://github.com/wordpress-mobile/WordPress-Android/commit/0ff109fba09b579df4f15b89dd7c951ed618f157)
7. [Fix base stats use case tests with advance until idle.](https://github.com/wordpress-mobile/WordPress-Android/commit/838b12c3d3246432f646c374524bea288c3702c0)
8. [Fix view model start/invoke tests with advance until idle.](https://github.com/wordpress-mobile/WordPress-Android/commit/4cb13c7a6c57f72b2118a65f1459ae55a9c6e0ff)

Fix Tests but also Ignore Changes:

1. [Fix site preview view model tests by triggering flow completion.](https://github.com/wordpress-mobile/WordPress-Android/commit/f8f18bd25bd07ab342910848514e598ba8e2946a)
2. [Fix reader related tests on pause/resume dispatcher failures.](https://github.com/wordpress-mobile/WordPress-Android/commit/ec3265f57d847c03859eaf377d7f3c19e2381d12)

Refactor Changes:

1. [Replace coroutines test rule test scope with helper function.](https://github.com/wordpress-mobile/WordPress-Android/commit/ede08a770b9380b2f1fa7b04b019ddb0ceb61d4c)
10. [Replace coroutines test rule test dispatcher with helper function.](https://github.com/wordpress-mobile/WordPress-Android/commit/90f5fab5355c17a87c38f4537596965e384fada1)

Minor but Mass Changes:

1. [Remove unused org wordpress android test import.](https://github.com/wordpress-mobile/WordPress-Android/commit/ee2a3bb451e4d54c44704a48aa1f73be491ec802) (`CoroutinesUtils.test { ... }` replacement/removal)
11. [Add missing experimental coroutines api annotation to all tests.](https://github.com/wordpress-mobile/WordPress-Android/commit/134a419e683c59dbe353aa0fc307198db8571ba6)

Minor Changes:

1. [Migrate test coroutine scope to test scope for qr code auth vm.](https://github.com/wordpress-mobile/WordPress-Android/commit/66126d3407e641f18c203e69f8ba7847a7d41016)
3. [Reformat base unit test class.](https://github.com/wordpress-mobile/WordPress-Android/commit/c4b6d316fa1bc591027990cac1bd1227f041b354)

-----

PS: @RenanLukas I added you as the main reviewer, that is, in addition to the @wordpress-mobile/apps-infrastructure team itself, but randomly, since I just wanted someone from the WordPress team to be aware of and sign-off on that change for WPAndroid.

-----

To test:

1. See the dependency tree diff result and verify correctness.
12. Thoroughly smoke test both, the WordPress and Jetpack apps, and see if they both work as expected.

Note the fact that:

- This `kotlinxCoroutinesVersion` update is only related to testing. As such, there was no code changes related to `main` code. However, that doesn't mean that the main apps' functionality couldn't have been affected by this update. So, please be extra careful when reviewing this PR and make sure to test both apps thoroughly.


-----

### Merge instructions

- [x] Wait for this PR to be fully review and approved.
- [x] Migrate any new tests that might have been added to `trunk` during the review period.

-----

## Regression Notes
1. Potential unintended areas of impact

    - Potential breakage of screens and core functionality as `org.jetbrains.kotlinx` is a library that is being added as a transitive dependency across lots of `androidx` and `com.google` and other libraries.
    - Some of the transitive dependencies added might be causing some kind of misbehaviour.

8. What I did to test those areas of impact (or what existing automated tests I relied on)

    - See `To test` section above.

13. What automated tests I added (or what prevented me from doing so)

    - `N/A`

-----

PR submission checklist:
